lang/LangPrimSource/PyrListPrim.cpp

```diff
@@ -492,41 +492,98 @@ int prEvent_Delta(struct VMGlobals *g, int numArgsPushed)
 	PyrSlot *a, key, dur, stretch, delta;
 	double fdur, fstretch;
 	int err;
+	PyrClass *restClass = s_rest->u.classobj;
+	PyrSlot *slot;
 
 	a = g->sp;  // dict
 
 	SetSymbol(&key, s_delta);
 	identDict_lookup(slotRawObject(a), &key, calcHash(&key), &delta);
 
 	if (NotNil(&delta)) {
-		slotCopy(a,&delta);
+		if (isKindOfSlot(&delta, restClass)) {
+			slot = slotRawObject(&delta)->slots;
+			err = slotDoubleVal(slot, &fdur);
+		} else {
+			err = slotDoubleVal(&delta, &fdur);
+		};
+		if (err) {
+			return err;
+		} else {
+			SetFloat(g->sp, fdur);
+			return errNone;
+		}
 	} else {
 		SetSymbol(&key, s_dur);
 		identDict_lookup(slotRawObject(a), &key, calcHash(&key), &dur);
-
 		err = slotDoubleVal(&dur, &fdur);
 		if (err) {
-			if (NotNil(&dur)) return err;
-			SetNil(a);
-			return errNone;
+			if (NotNil(&dur)) {
```

Comment:
At line 520, there are three 'if's in a row. I'm unclear which ones you would like to combine.

But I don't think it matters, because I'm quite sure they can't be combined.

The current logic tree is (and I'll invert it because the else branches are simpler, keeping the branches closer to the conditions):

- Did float conversion fail?
  - False(else): Use the converted float.
  - True(then): Is it nil?
    - False: Primitive returns with nil.
    - True: Is it a Rest?
      - False: Wrong type.
      - True: Get the float. Did this float conversion fail?
        - False: Use the converted float.
        - True: The Rest contains something other than a number -- wrong type.

If you're thinking `if(err && NotNil(&dur))` then you would have to collapse "False(else): Use the converted float" and "False: Primitive returns with nil" into one 'else' branch, which you can't do.

If you're thinking `if(NotNil(&dur) && isKindOfSlot(&dur, restClass)`, then you would have to collapse "False: Primitive returns with nil" and "False: Wrong type" into one 'else' branch, which you can't do.

If you're thinking `if(err && isKindOfSlot(&dur, restClass))`, then you would lose the fact that a `nil` dur should *not* throw an error.